libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h

```diff
@@ -0,0 +1,525 @@
+};
+
+template<_CUDA_VSTD::size_t _Alignment>
+struct aligned_size_t {
+    static constexpr _CUDA_VSTD::size_t align = _Alignment;
+    _CUDA_VSTD::size_t value;
+    _LIBCUDACXX_INLINE_VISIBILITY
+    explicit aligned_size_t(size_t __s) : value(__s) { }
+    _LIBCUDACXX_INLINE_VISIBILITY
+    operator size_t() const { return value; }
+};
+
+// Type only used for logging purpose
+enum async_contract_fulfillment {
+    none,
+    async
+};
+
+enum class __space {
+    __local,
+    __shared,
+    __cluster,
+    __global,
+    __constant,
+    __grid_constant
+};
+
+template<__space _Sp>
+using __space_constant = _CUDA_VSTD::integral_constant<__space, _Sp>;
+
+_LIBCUDACXX_DEVICE
+bool __is_cluster_shared(const void * __p) {
+    NV_IF_TARGET(NV_PROVIDES_SM_90,
+        (return __isClusterShared(__p);),
+        (return false;)
+    )
+}
+
+_LIBCUDACXX_DEVICE
+bool __is_grid_constant(const void * __p) {
+#ifdef _LIBCUDACXX_COMPILER_NVCC_BELOW_11_7
+    return false;
+#else
+    return __isGridConstant(__p);
+#endif
+}
+
+#define _LIBCUDACXX_HANDLE_SPACE(intrinsic, enum_v, name, ...) \
+    if (intrinsic(name)) { \
+        using name ## _space_t = __space_constant<__space::enum_v>; \
+        __VA_ARGS__ \
+    } else
+
+#define _LIBCUDACXX_HANDLE_POINTER_SPACE(name, ...) \
+    NV_IF_ELSE_TARGET( \
+        NV_IS_DEVICE, ( \
+            _LIBCUDACXX_HANDLE_SPACE(__is_cluster_shared,__cluster, name, __VA_ARGS__) \
+            _LIBCUDACXX_HANDLE_SPACE(__isShared, __shared, name, __VA_ARGS__) \
+            _LIBCUDACXX_HANDLE_SPACE(__isGlobal, __global, name, __VA_ARGS__) \
+            _LIBCUDACXX_HANDLE_SPACE(__is_grid_constant, __grid_constant, name, __VA_ARGS__) \
+            _LIBCUDACXX_HANDLE_SPACE(__isConstant, __constant, name, __VA_ARGS__) \
+            _LIBCUDACXX_HANDLE_SPACE(/* __isLocal when not broken */ bool, __local, name, __VA_ARGS__) \
+            {} \
+        ), ( \
+            _LIBCUDACXX_HANDLE_SPACE(bool, __global, name, __VA_ARGS__) \
+            {} \
+        ) \
+    )
+
+template<typename _Tag, _CUDA_VSTD::size_t _Value>
+struct __down_convertible_constant {
+    template<_CUDA_VSTD::size_t _OtherValue, typename = _CUDA_VSTD::__enable_if_t<(_OtherValue < _Value)>>
+    _LIBCUDACXX_INLINE_VISIBILITY
+    constexpr operator __down_convertible_constant<_Tag, _OtherValue>() const {
+        return {};
+    }
+
+    static const constexpr _CUDA_VSTD::size_t value = _Value;
+};
+
+struct __alignment_tag {};
+template<_CUDA_VSTD::size_t _Alignment>
+using __alignment = __down_convertible_constant<__alignment_tag, _Alignment>;
+
+template<_CUDA_VSTD::size_t _GuaranteedAlignment,
+    _CUDA_VSTD::size_t _MaxInterestingAlignment,
+    _CUDA_VSTD::size_t _MinInterestingAlignment,
+    _CUDA_VSTD::size_t _Alignment,
+    typename = void>
+struct __invoke_if_applicable {
+    template<typename _Fn>
+    _LIBCUDACXX_INLINE_VISIBILITY static async_contract_fulfillment __invoke(_Fn && __f) {
```

Comment:
As a reader the static and the return type are the most important parts here, so please add a line break after the macro
```suggestion
    _LIBCUDACXX_INLINE_VISIBILITY
    static async_contract_fulfillment __invoke(_Fn && __f) {
```